register-machines/src/main.rs

```diff
@@ -13,66 +15,165 @@ enum Instruction {
+                }
+            },
+            Halt => { break; }
+        }
+    }
+    (l, rs)
+}
+fn encode_instruction(ins: &Instruction) -> Godel {
+    match ins {
+        Add(i, j) => 
+            encode_pair1(&Godel::from(2 * i), &Godel::from(*j)),
+        Sub(i, j, k) => 
+            encode_pair1(&Godel::from(2 * i + 1),
+            &encode_pair2(&Godel::from(*j), &Godel::from(*k))
+        ),
+        Halt => Godel::ZERO
+    }
 }
 // <<x,y>> = (2^x)*(2y+1)
-fn encode_pair1(x: u128, y: u128) -> u128 {
-    unimplemented!();
+fn encode_pair1(x: &Godel, y: &Godel) -> Godel {
+    (Godel::from(2) * y + 1) << x.to_u32_wrapping()
 }
 // <x,y> = (2^x)*(2y+1)-1
-fn encode_pair2(x: u128, y: u128) -> u128 {
-    unimplemented!();
+fn encode_pair2(x: &Godel, y: &Godel) -> Godel {
+    encode_pair1(x, y) - 1
 }
-fn encode_list_to_godel(l: &[u128]) -> u128 {
-    unimplemented!();
+fn encode_list_to_godel(l: &[Godel]) -> Godel {
+    if l.is_empty() { return Godel::ZERO; } 
+    encode_pair1(&l[0], &encode_list_to_godel(&l[1..])) 
 }
-fn encode_program_to_list(program: &[Instruction]) -> Vec<u128> {
-    unimplemented!();
+fn encode_program_to_list(program: &[Instruction]) -> Vec<Godel> {
+    program.iter().map(|x| encode_instruction(x)).collect()
+}
+// Returns 0 if there are no trailing zeros, else return trailing zero count (in binary)
+fn trailing_zeros_in_binary(x: &Godel) -> Godel {
+    let mut b = x.clone(); 
+    let mut c = Godel::new();
+    if b != 0 {
+        b = (b.clone() ^ (b - 1)) >> 1; 
+        while b != 0 {
+            c += 1;
+            b >>= 1;
+        }
+    }
+    c
+}
+fn decode_instruction(ins: &Godel) -> Instruction {
+    if *ins == 0 { return Halt; }
+    let (x, y) = decode_pair1(ins); 
+    let i: Godel = x.clone() / 2;
+    if x % 2 != 0 {
+        let (j, k) = decode_pair2(&y);
+        Sub(i.try_into().unwrap(), j.try_into().unwrap(), k.try_into().unwrap())
+    } else {
+        Add(i.try_into().unwrap(), y.try_into().unwrap())
+    }
 }
 // a = (2^x)*(2y+1)
-fn decode_pair1(a: u128) -> (u128, u128) {
-    unimplemented!();
+fn decode_pair1(a: &Godel) -> (Godel, Godel) {
+    let x: Godel = trailing_zeros_in_binary(a); 
+    let z = Godel::from(a >> x.to_u32_wrapping());
+    let y: Godel = (z - 1) / 2;
+    (x, y)
 }
 // a = (2^x)*(2y+1)-1
-fn decode_pair2(a: u128) -> (u128, u128) {
-    unimplemented!();
+fn decode_pair2(a: &Godel) -> (Godel, Godel) {
+    decode_pair1(&Godel::from(a + 1))
 }
-fn decode_godel_to_list(g: u128) -> Vec<u128> {
-    unimplemented!();
+fn decode_godel_to_list(g: &Godel) -> Vec<Godel> {
+    if *g == 0 { return Vec::new(); }
+    let (x, xs) = decode_pair1(g);
+    let mut gs = vec![x]; 
+    gs.splice(gs.len().., decode_godel_to_list(&xs));
+    gs 
 }
-fn decode_list_to_program(program: &[u128]) -> Vec<Instruction> {
-    unimplemented!();
+fn decode_list_to_program(program: &[Godel]) -> Vec<Instruction> {
+    program.iter().map(|x| decode_instruction(x)).collect()
 }
```

Comment:
This is great, I love it. Only comment is that you can map single-argument functions without closures:
`program.iter().map(decode_instruction).collect()